### ISSUE TITLE
This example does not compile under Solaris with Oracle Developer Studio 12.6

### ISSUE THREAD

**Issue by stegua** [r-pkg-examples/rcpp-modules-student#2]
This example does not compile on [R-hub](https://builder.r-hub.io/) using the platform option 

- *Oracle Solaris 10, x86, 32 bit, R-release, Oracle Developer Studio 12.6* 

After updating the package obtained running the command:

`R CMD build .`

the system reports the following error:

`Error: package or namespace load failed for â€˜RcppStudentâ€™ in .doLoadActions(where, attach):
 error in load action .__A__.1 for package RcppStudent: loadModule(module = "RcppStudentEx", what = TRUE, env = ns, loadNow = TRUE): Unable to load module "RcppStudentEx": negative length vectors are not allowed
Error: loading failed
Execution halted`

Thanks.

**Comment by coatless** [comment]
@stegua thanks for the ping. 

Running `devtools::check_rhub(platform ="solaris-x86-patched-ods")` does indeed give an error on Solaris. I haven't the slightest clue why it is showing: 

>  error in load action .__A__.1 for package RcppStudent: loadModule(module = "RcppStudentEx", what = TRUE, env = ns, loadNow = TRUE): Unable to load module "RcppStudentEx": negative length vectors are not allowed

None of the vectors are "negative length" or are large enough not to qualify as an integer. 

With that being said, running the non-ods Solaris image, `devtools::check_rhub(platform ="solaris-x86-patched")`, yields a clean build a bit with some configuration warnings on the image machine itself. 

So, I don't think the current package structure would be problematic if I placed the example on CRAN. I say this since none of the other images on Rhub (Windows, macOS, ubuntu) are erroring and it conforms to how the test code for the Rcpp Module feature in Rcpp is organized.

https://github.com/RcppCore/Rcpp/blob/27f24f09bfaebc0455fd8e88d8c8abe37fb76c6f/inst/tinytest/testRcppModule/src/rcpp_module.cpp#L46-L72


<details>
<summary> R CMD check output from `devtools::check_rhub(platform ="solaris-x86-patched")` </summary>

```
* checking top-level files ... NOTE
Files ‘README.md’ or ‘NEWS.md’ cannot be checked without ‘pandoc’ being installed.
* checking compilation flags used ... NOTE
Compilation used the following non-portable flag(s):
  ‘-march=pentiumpro’
```
</details>

<details>
<summary> Results from `RcppStudent.Rcheck/00install.out` obtained with `devtools::check_rhub(platform ="solaris-x86-patched-ods")`</summary>

```cpp
** using staged installation
** libs
/opt/developerstudio12.6/bin/CC -std=c++11 -library=stdcpp,CrunG3  -I"/opt/R/R-4.0.5/lib/R/include" -DNDEBUG  -I'/export/home/XA9Ypka/R/Rcpp/include' -I/opt/csw/include   -KPIC  -O -xlibmil -xtarget=native -xcache=generic -nofstore  -c RcppExports.cpp -o RcppExports.o
/opt/developerstudio12.6/bin/CC -std=c++11 -library=stdcpp,CrunG3  -I"/opt/R/R-4.0.5/lib/R/include" -DNDEBUG  -I'/export/home/XA9Ypka/R/Rcpp/include' -I/opt/csw/include   -KPIC  -O -xlibmil -xtarget=native -xcache=generic -nofstore  -c student.cpp -o student.o
/opt/developerstudio12.6/bin/CC -std=c++11 -library=stdcpp,CrunG3  -I"/opt/R/R-4.0.5/lib/R/include" -DNDEBUG  -I'/export/home/XA9Ypka/R/Rcpp/include' -I/opt/csw/include   -KPIC  -O -xlibmil -xtarget=native -xcache=generic -nofstore  -c student_export.cpp -o student_export.o
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/Module_generated_class_constructor.h", line 81: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 10:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::constructor<std::string, int, bool>(const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 10:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 100: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 10:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::constructor<std::string, int, bool>(const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 10:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/Module_generated_method.h", line 28: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 11:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::method<std::string>(const char*, std::string (Student::*)(), const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 11:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 239: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 11:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::method<std::string>(const char*, std::string (Student::*)(), const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 11:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/Module_generated_method.h", line 28: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 12:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::method<int>(const char*, int(Student::*)(), const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 12:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/Module_generated_method.h", line 28: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 13:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::method<bool>(const char*, bool(Student::*)(), const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 13:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/Module_generated_method.h", line 28: Warning: docstring hides Rcpp::class_Base::docstring.
"student_export.cpp", line 14:     Where: While instantiating "Rcpp::class_<Student>& Rcpp::class_<Student>::method<std::vector<int>>(const char*, std::vector<int>(Student::*)(), const char*, bool(*)(SEXPREC**,int))".
"student_export.cpp", line 14:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 288: Warning: name hides Rcpp::class_Base::name.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 276:     Where: While instantiating "Rcpp::class_<Student>::method_names()".
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 276:     Where: Instantiated from Rcpp::class_<Student>::class_(const char*, const char*).
"student_export.cpp", line 9:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 309: Warning: name hides Rcpp::class_Base::name.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 296:     Where: While instantiating "Rcpp::class_<Student>::methods_arity()".
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 296:     Where: Instantiated from Rcpp::class_<Student>::class_(const char*, const char*).
"student_export.cpp", line 9:     Where: Instantiated from non-template code.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 333: Warning: name hides Rcpp::class_Base::name.
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 320:     Where: While instantiating "Rcpp::class_<Student>::methods_voidness()".
"/export/home/XA9Ypka/R/Rcpp/include/Rcpp/module/class.h", line 320:     Where: Instantiated from Rcpp::class_<Student>::class_(const char*, const char*).
"student_export.cpp", line 9:     Where: Instantiated from non-template code.
10 Warning(s) detected.
/opt/developerstudio12.6/bin/CC -std=c++11 -library=stdcpp,CrunG3  -I"/opt/R/R-4.0.5/lib/R/include" -DNDEBUG  -I'/export/home/XA9Ypka/R/Rcpp/include' -I/opt/csw/include   -KPIC  -O -xlibmil -xtarget=native -xcache=generic -nofstore  -c student_non_modules.cpp -o student_non_modules.o
/opt/developerstudio12.6/bin/CC -std=c++11 -library=stdcpp,CrunG3 -G -L/opt/csw/lib -o RcppStudent.so RcppExports.o student.o student_export.o student_non_modules.o
installing to /export/home/XA9Ypka/R/00LOCK-RcppStudent/00new/RcppStudent/libs
** R
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for â€˜RcppStudentâ€™ in .doLoadActions(where, attach):
 error in load action .__A__.1 for package RcppStudent: loadModule(module = "RcppStudentEx", what = TRUE, env = ns, loadNow = TRUE): Unable to load module "RcppStudentEx": negative length vectors are not allowed
Error: loading failed
Execution halted
ERROR: loading failed
* removing â€˜/export/home/XA9Ypka/R/RcppStudentâ€™
* restoring previous â€˜/export/home/XA9Ypka/R/RcppStudentâ€™
```

</details>

**Comment by stegua** [comment]
@coatless Thanks for the quick reply!

I confirm that on the standard Solaris image with *gcc*, we don't get the strange error.

However, I am trying to figure out why my package was rejected by CRAN for a Solaris issue, as reported in the following question:

- [R Cran package with Rcpp builds on every system but Solaris](https://stackoverflow.com/questions/67230744/r-cran-package-with-rcpp-builds-on-every-system-but-solaris)

**Comment by coatless** [comment]
@stegua I'll take a closer look probably toward the end of the week. The term is ending here so my time is sparse.

In the interim, maybe try to setup a local solaris build:

Jeroen created a solaris image long ago at: https://github.com/jeroen/solarisvm

However, it seems like a better resource would be setting up an image following rhub guides:

 - The R-hub Solaris vmware/virtualbox images: https://files.r-hub.io/solaris/
 - Using the VM: https://github.com/r-hub/solarischeck/tree/master/packer#using-the-vm
 - How to create a Solaris VM yourself: https://docs.r-hub.io/technical/solaris/

**Comment by stegua** [comment]
This morning I successfully compiled and executed a standalone version of my code on Solaris 11 via Virtual Box, but I had troubles with r_base. At least, I have reduced the scope of code where to look for an error, that is, in the wrapper to R.

Now, I am trying to downgrade to Solaris 10 following the [solarischeck](https://github.com/r-hub/solarischeck/tree/master/packer#using-the-vm) instructions.

I will keep you informed about any progress.

Thanks for your support.
